### PR TITLE
Fixes #13703 - added db info into foreman-debug

### PR DESCRIPTION
**Pull request by lzap**
This patch adds more information from our database into debug tarball. So far
we have been only shipping settings and auth_sources.

This is selection of our tables which I think might be useful in giving
support. All the data that can be sensitive are hashed via plain md5 function
(no salt). The idea is to make them unreadable. This is not top-security, but
debug tarballs are not usually disclosed publicly. It also allows to check
things in database when there are hostnames or logins mentioned in
ticket/bugzilla.

Here is an example output: https://gist.github.com/lzap/50a3760148413496e4ba

I would like also to export all templates in separate files once I find the correct PgSQL command: http://stackoverflow.com/questions/35409314/dump-individual-text-fields-into-separate-files-in-psql

### CODE REVIEW

**Comment by lzap** [comment]
Useful link for testing this PR:

```
https://raw.githubusercontent.com/lzap/foreman/debug-db-13703/script/foreman-debug
```

File: script/foreman-debug [hunk 1]
```diff
@@ -293,9 +303,24 @@ add_files /etc/{httpd,apache2}/conf.d/*
 add_files /etc/{httpd,apache2}/conf.d/*/*
 add_files /var/log/{httpd,apache2}/*error_log*
 add_files /var/log/{httpd,apache2}/foreman-ssl_access_ssl.log*
-add_cmd "echo \"select id,name,value from settings where name not similar to '%(pass|key|secret)'\" | su postgres -c 'psql foreman'" "foreman_settings_table"
-add_cmd "echo 'select type,name,host,port,account,base_dn,attr_login,onthefly_register,tls from auth_sources' | su postgres -c 'psql foreman'" "foreman_auth_table"
 add_cmd "foreman-selinux-relabel -nv" "foreman_filecontexts"
+add_sql "select id,name,value from settings where name not similar to '%(pass|key|secret)' order by name" settings
```
Comment (ohadlevy) [inline script/foreman-debug]: maybe only when value != default_value?

Comment (lzap) [inline script/foreman-debug]: Well, value seems to be blank when it's default, see the gist.

File: script/foreman-debug [hunk 2]
```diff
@@ -79,6 +79,15 @@ add_cmd() {
   eval $CMD >> "$DIR/$OUT" 2>&1
 }
 
+# add outout of the sql and redirect possible errors there
+add_sql() {
+  SQL=$1
+  OUT=$2
+  printv " - $OUT"
+  echo -e "SQL> $SQL\n" > "$DIR/db_$OUT"
+  echo "$SQL" | su postgres -c 'psql foreman' >> "$DIR/db_$OUT" 2>&1
```
Comment (domcleal) [inline script/foreman-debug]: This seems to assume a certain DB type, and username - it should be driven from the env settings?

File: script/foreman-debug [hunk 3]
```diff
@@ -293,9 +303,24 @@ add_files /etc/{httpd,apache2}/conf.d/*
...
+add_sql "select id,md5(name) as name,managed,build,organization_id,location_id from hosts order by name" hosts
+add_sql "select id,name,organization_id,location_id,vm_defaults from hostgroups order by name" hostgroups
+add_sql "select id,name,path,media_path,config_path,image_path,os_family from media order by name" media
+add_sql "select id,host_id,mac,ip,type,name,subnet_id,domain_id from nics order by host_id,id" nics
+add_sql "select id,name,major,minor,type,description from operatingsystems order by name" operatingsystems
+add_sql "select id,name,url from smart_proxies order by name" smart_proxies
+add_sql "select id,name,network,mask,dhcp_id,tftp_id,dns_id from subnets order by name" subnets
+add_sql "select id,name,type from taxonomies order by name" taxonomies
+add_sql "select id,name from config_templates order by name" config_templates
+add_sql "select 2,md5(value) as value,expires,host_id from tokens order by expires" tokens
```
Comment (domcleal) [inline script/foreman-debug]: What's the `2` for?

File: script/foreman-debug [hunk 4]
```diff
@@ -293,9 +303,24 @@ add_files /etc/{httpd,apache2}/conf.d/*
...
+add_sql "select id,name,hosts_count from environments order by name" environments
+add_sql "select id,md5(name) as name,managed,build,organization_id,location_id from hosts order by name" hosts
+add_sql "select id,name,organization_id,location_id,vm_defaults from hostgroups order by name" hostgroups
+add_sql "select id,name,path,media_path,config_path,image_path,os_family from media order by name" media
+add_sql "select id,host_id,mac,ip,type,name,subnet_id,domain_id from nics order by host_id,id" nics
+add_sql "select id,name,major,minor,type,description from operatingsystems order by name" operatingsystems
+add_sql "select id,name,url from smart_proxies order by name" smart_proxies
+add_sql "select id,name,network,mask,dhcp_id,tftp_id,dns_id from subnets order by name" subnets
+add_sql "select id,name,type from taxonomies order by name" taxonomies
+add_sql "select id,name from config_templates order by name" config_templates
```
Comment (domcleal) [inline script/foreman-debug]: no such table?

File: script/foreman-debug [hunk 5]
```diff
@@ -293,9 +303,24 @@ add_files /etc/{httpd,apache2}/conf.d/*
...
-add_cmd "echo \"select id,name,value from settings where name not similar to '%(pass|key|secret)'\" | su postgres -c 'psql foreman'" "foreman_settings_table"
-add_cmd "echo 'select type,name,host,port,account,base_dn,attr_login,onthefly_register,tls from auth_sources' | su postgres -c 'psql foreman'" "foreman_auth_table"
 add_cmd "foreman-selinux-relabel -nv" "foreman_filecontexts"
+add_sql "select id,name,value from settings where name not similar to '%(pass|key|secret)' order by name" settings
+add_sql "select type,name,host,port,account,base_dn,attr_login,onthefly_register,tls from auth_sources order by name" auth_sources
+add_sql "select id,name,type from compute_resources order by name" compute_resources
+add_sql "select id,name,fullname,dns_id from domains order by name" domains
+add_sql "select id,name,hosts_count from environments order by name" environments
+add_sql "select id,md5(name) as name,managed,build,organization_id,location_id from hosts order by name" hosts
+add_sql "select id,name,organization_id,location_id,vm_defaults from hostgroups order by name" hostgroups
```
Comment (domcleal) [inline script/foreman-debug]: host groups don't have org/loc_id fields

**Comment by domcleal** [comment]
This is a huge amount of quite sensitive information about people's environments, especially lists of hosts and NICs.  It really must be off by default I think.

I think it would be better to use the existing database export tools and perhaps implement #11176 to "clean" some sensitive information rather than to hardcode lots of columns and tables, then dumping this data into foreman-debug.

File: script/foreman-debug (hunk 2 above)
Comment (lzap) [inline script/foreman-debug]: Yes, this is PostgreSQL only, I will add a check. Username is not relevant tho, it's actually database name here, we are accessing this from the superuser.

File: script/foreman-debug (hunk 3 above)
Comment (lzap) [inline script/foreman-debug]: Typo, thanks.

File: script/foreman-debug (hunk 4 above)
Comment (lzap) [inline script/foreman-debug]: Oh thanks, was testing this on Satellite.

File: script/foreman-debug (hunk 5 above)
Comment (lzap) [inline script/foreman-debug]: Thanks, missed that.

**Comment by lzap** [comment]
Ok I agree that's not good and I kinda expected that. Removing the MD5 sums and adding new option `-E` that is explicitly needed to create what I call _Detailed report_. Pushed, fixed all the comments. I hope to hit 1.11 with this.

File: script/foreman-debug [hunk 6]
```diff
@@ -303,6 +329,28 @@ add_files /etc/sysconfig/pgsql
...
+  add_sql "select id,host_id,mac,ip,type,name,subnet_id,domain_id from nics order by host_id,id" nics
+  add_sql "select id,name,major,minor,type,description from operatingsystems order by name" operatingsystems
+  add_sql "select id,name,url from smart_proxies order by name" smart_proxies
+  add_sql "select id,name,network,mask,dhcp_id,tftp_id,dns_id from subnets order by name" subnets
+  add_sql "select id,name,type from taxonomies order by name" taxonomies
+  add_sql "select id,name from provisioning_templates order by name" provisioning_templates
+  add_sql "select id,value as value,expires,host_id from tokens order by expires" tokens
+  add_sql "select id,login as login,admin,auth_source_id from users order by id" users
+  add_sql "select id,name from roles order by name" roles
+  add_sql "copy provisioning_templates(id,name,template) TO '$DIR/db_provisioning_templates.csv' (format csv)" provisioning_templates_export
```
Comment (lzap) [inline script/foreman-debug:385]: For the record, the last command will fail on MySQL adapter, but it wont error out, the output will only contain an error message instead of the data. All the others should work. @ohadlevy is currently helping me to test this on MySQL instance.

**Comment by lzap** [comment]
Also changed the SCL to 'tfm'.

**Comment by lzap** [comment]
I also think it is worth moving Apache `access_log(s)` in the detailed section - also contains quite sensitive data (client IPs). Any other tips for this move?

**Comment by lzap** [comment]
Or is there something we are not reporting for security reasons we could do with the explicit option now?

File: script/foreman-debug [hunk 7]
```diff
@@ -2,7 +2,7 @@
 # vim:sw=2:ts=2:et
 
 export LC_ALL=C
-export SCLNAME=ruby193
+export SCLNAME=tfm
```
Comment (domcleal) [inline script/foreman-debug:5]: This is unrelated, please put it into a different ticket and PR.

File: script/foreman-debug [hunk 8]
```diff
@@ -19,6 +19,8 @@ This program can be used on Foreman instances, Smart Proxy instances or
 backend services separately.
 
 OPTIONS:
+  -E      Detailed output incl. host names or domains
```
Comment (domcleal) [inline script/foreman-debug]: This is misleading, it's the contents of the database and implies that the hostname and domain will be obscured otherwise.

File: script/foreman-debug [hunk 9]
```diff
@@ -79,6 +83,21 @@ add_cmd() {
   eval $CMD >> "$DIR/$OUT" 2>&1
 }
 
+# add outout of the sql and redirect possible errors there
+add_sql() {
+  SQL=$1
+  OUT=$2
+  printv " - $OUT"
+  echo -e "SQL> $SQL\n" > "$DIR/db_$OUT"
+  if [ "$DB_ADAPTER" = "postgresql" ]; then
```
Comment (domcleal) [inline script/foreman-debug:99]: This should be replaced by a database dump I think, per my earlier comment.  The code for this is already in Foreman and it handles environment configuration far better, plus avoids the pile of SQL in this shell script.

File: script/foreman-debug [hunk 10]
```diff
@@ -19,6 +19,8 @@ This program can be used on Foreman instances, Smart Proxy instances or
 backend services separately.
 
 OPTIONS:
+  -E      Detailed output incl. host names or domains
+          (THIS OPTION ADDS MORE SENSITIVE DATA IN THE REPORT)
```
Comment (domcleal) [inline script/foreman-debug]: This is also somewhat misleading, the flag should be more specific to say that it's including contents of the database.

**Comment by domcleal** [comment]
New flag(s) and behaviour need to be in the man page too.

File: script/foreman-debug (hunk 7 above)
Comment (lzap) [inline script/foreman-debug:5]: http://projects.theforeman.org/issues/13803

File: script/foreman-debug (hunk 10 above)
Comment (lzap) [inline script/foreman-debug]: Do you have a particular example how the flag should look like?

I do want to export only particular items via SQL. Your idea to provide full database dump is good, but I want to implement this in a separate flag, because the dump can very quickly run out of size bounds.

File: script/foreman-debug (hunk 9 above)
Comment (lzap) [inline script/foreman-debug:99]: Database dump is a good idea, but I will implement that separately for the reasons above.

**Comment by lzap** [comment]
Added three new explicit flags and man texts.

File: script/foreman-debug (hunk 10 above)
Comment (domcleal) [inline script/foreman-debug]: What size difference is there between the selective tables and dumping the database?  You've listed some potentially large tables anyway, there don't seem to be many in my opinion that would be any larger than say, `hosts` or `nics`.

_Edit: please reply in main thread, not here._

**Comment by domcleal** [comment]
Moving my comment out of the outdated diff thread:

> I do want to export only particular items via SQL. Your idea to provide full database dump is good, but I want to implement this in a separate flag, because the dump can very quickly run out of size bounds.

What size difference is there between the selective tables and dumping the database? You've listed some potentially large tables anyway, there don't seem to be many in my opinion that would be any larger than say, hosts or nics.

**Comment by lzap** [comment]
> What size difference is there between the selective tables and dumping the database? You've listed some potentially large tables anyway, there don't seem to be many in my opinion that would be any larger than say, hosts or nics.

Well, I haven't listed audits, fact_values or reports, these are my biggest concern.

## 

Later,
 Lukas #lzap Zapletal

**Comment by lzap** [comment]
Also, biggest advantage of SQL approach is having two levels of details. Full database dump gives you everything (including passwords - in encrypted form) while SQL approach is less verbose and does not include any passwords.

File: script/foreman-debug [hunk 11]
```diff
@@ -303,6 +354,42 @@ add_files /etc/sysconfig/pgsql
 add_files "/var/lib/pgsql/data/pg_log/*"
 add_cmd "foreman-rake plugin:list" "plugin_list"
 
+if [ $DB_DUMP -eq 1 ]; then
+  if [ "$DB_ADAPTER" = "postgresql" ]; then
+    printv "Making full PostgreSQL database dump"
+    su postgres -c "pg_dump -Fc $DB_NAME" > "$DIR/database_pg.dump"
```
Comment (domcleal) [inline script/foreman-debug:360]: Use the existing db:dump rake tasks for this, no need for detecting adapters etc.

File: man/foreman-debug.8.asciidoc [hunk 12]
```diff
@@ -39,17 +39,55 @@ OPTIONS
...
+
+By default, foreman-debug does not include any data that are sensitive, such as
+passwords, tokens, host names or IP addresses. It is possible tho to explicitly
+include such information via options:
+
+Database tables (--db-tables)
+
+Adds selection of tables from the main database,  most importantly:
+
+* Hosts and Hostgroups
```
Comment (domcleal) [inline man/foreman-debug.8.asciidoc]: "host groups"

File: man/foreman-debug.8.asciidoc [hunk 13]
```diff
@@ -39,17 +39,55 @@ OPTIONS
...
+* Subnets and Domains
+* Operating systems and Media
+* Provisioning templates
+* Taxonomies (organization and location names)
```
Comment (domcleal) [inline man/foreman-debug.8.asciidoc]: Don't use "Taxonomies" in user docs, just "organizations and locations"

File: man/foreman-debug.8.asciidoc [hunk 14]
```diff
@@ -39,17 +39,55 @@ OPTIONS
...
   -q      Quiet mode
   -v      Verbose mode
   -u      Upload tarball
   -h      Shows this message
 
+SECURITY
+--------
+
+By default, foreman-debug does not include any data that are sensitive, such as
+passwords, tokens, host names or IP addresses. It is possible tho to explicitly
```
Comment (domcleal) [inline man/foreman-debug.8.asciidoc]: 1) "possible _though_ to"

2) Is this true?  By default it includes production.log, which contains IP addresses and hostnames, so I wouldn't claim this.

**Comment by domcleal** [comment]
> Well, I haven't listed audits, fact_values or reports, these are my biggest concern.

Good point, those should probably be excluded.

> Also, biggest advantage of SQL approach is having two levels of details. Full database dump gives you everything (including passwords - in encrypted form) while SQL approach is less verbose and does not include any passwords.

Yes, that makes sense.  At the moment the implementation needs to change, to move these SQL statements out of this file.  I don't think it's the right place for them as they won't get maintained in line with database changes.  They also miss quite a few useful columns and/or tables, making it an arbitrary and selective list now.

It seems preferable to implement this as a rake task, similar to prod2dev that can simply iterate over known tables and print the output from all columns.  You can identify and exclude encrypted columns easily (the Encryptable concern is used and knows them) and skip tables that aren't needed.  The rake task can also be tested from test/lib/tasks/.

File: man/foreman-debug.8.asciidoc (hunk 14 above)
Comment (lzap) [inline man/foreman-debug.8.asciidoc]: Reworded whole part.

File: man/foreman-debug.8.asciidoc (hunk 13 above)
Comment (lzap) [inline man/foreman-debug.8.asciidoc]: Fixed both.

File: script/foreman-debug (hunk 11 above)
Comment (lzap) [inline script/foreman-debug:360]: Large deployments of Foreman (e.g. Katello and other plugins) bootup times are very slow, foreman-debug is already quite slow today and I don't want to boot the whole stack just because of this. Full dump is indeed explicit and this could be fine, but the adapter/credentials detection is still needed for the SQLs below.

**Comment by lzap** [comment]
All failed commands (including the new SQL commands) silently fail putting the error message into the output file. In general, it is good idea to catch errors early, but this is a debug helper script. A non-core component which won't break if we decide to change the schema.

I will perhaps refactor this later.

File: script/foreman-debug (hunk 11 above)
Comment (domcleal) [inline script/foreman-debug:360]: This isn't a default option, so the speed makes little difference.  Using the existing rake task means there won't be a duplicated set of code, and in particular, the existing code handles things like remote databases and credentials that this code doesn't.

**Comment by domcleal** [comment]
I think you've hit precisely my point - having something that fails quietly is bad.  This script isn't well-maintained at the moment, and embedding a lot of schema information inside it will be a continuing headache.

**Comment by lzap** [comment]
I will return to this PR shortly, since we now have an exporter task which creates custom-made reports, I will go ahead and add exporters for most important data and use it in debug.
